### PR TITLE
[SPARK-38791][SQL] Output parameter values of error classes in the SQL style

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose new trait `QueryErrorsBase` which is supposed to be used by `Query.*Errors`, and new method `toSQLValue()`. The method converts a parameter value of error classes to its SQL representation.

### Why are the changes needed?
To improve user experience with Spark SQL. Users should see values in error messages in unified SQL style.

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
By running the modified test suites:
```
$ build/sbt "test:testOnly *QueryExecutionErrorsSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@ivoson @yutoacts @leesf @cloud-fan @gengliangwang Could you review this PR, please.

**Review by leesf — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.apache.spark.sql.catalyst.expressions.Literal
+import org.apache.spark.sql.types.StringType
+import org.apache.spark.unsafe.types.UTF8String
+
+trait QueryErrorsBase {
+  // Converts an error class parameter to its SQL representation
+  def toSQLValue(v: Any): String = v match {
+    case _: UTF8String => Literal.create(v, StringType).sql
+    case _ => Literal(v).sql
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala:28]: how can we distinguish binary type and array<byte> if the input is `Array[Byte]`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala:28]: maybe `toSQLValue` should take an optional data type parameter

Comment (ivoson) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala:28]: Checked the [Literal.sql](https://github.com/apache/spark/blob/v3.2.1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala#L454) function, for some data types the sql string may be not as expected to be  shown in error msg, for example `null` values and `FloatType` values, the sql string would be a `cast` expression.

```
    // Float type doesn't have a suffix
    case (v: Float, FloatType) =>
      val castedValue = v match {
        case _ if v.isNaN => "'NaN'"
        case Float.PositiveInfinity => "'Infinity'"
        case Float.NegativeInfinity => "'-Infinity'"
        case _ => s"'$v'"
      }
      s"CAST($castedValue AS ${FloatType.sql})"
```

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala:28]: > the sql string may be not as expected

What is the expected from your point of view? Output of error messages should be consistent with other places, if we output a cast in other places, ok, let's do that in errors as well.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala:28]: I agree it's weird to see an unrelated `CAST` in the error message. Can we add some special cases here? e.g. null value is `NULL`, float value should be the same as the double value.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala]: Added an alternative method ...

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala]: Extended the `toSQLValue` function to handle the special cases.

**Comment by MaxGekk** [comment]
@cloud-fan @ivoson Could you review this PR one more time, please.

**Comment by ivoson** [comment]
LGTM

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @ivoson @cloud-fan @leesf for review.

**Comment by cloud-fan** [comment]
shall we merge it to 3.3 as well? error message improvement is a safe change and we can keep backporting before rc1.

**Pull request closed without merging**